contracts/ExitHandler.sol

```diff
@@ -29,42 +29,392 @@ contract ExitHandler is DepositHandler {
+      TxLib.Input memory competingInput = competingTx.input[InputNo];
+
+      // check if same or not ?      
+      require(inOutput.owner == owner);
+      require(inOutput.amount == competingTx.outs[competingInput.outpoint.pos]);
+
+    } else {
+      // if someone challenged it before and became successful
+      LimboChallenge memory prevChallenge = limboExit.challenge;
+      bytes32[] memory _prevProof = prevChallenge.proof;
+      uint64 previousPos;
+      bytes32 previousHash;
+      bytes memory previousTxData;
+
+      (previousPos, previousHash, previousTxData) = TxLib.validateProof(32 * (_youngestInputProof.length + 2) + 64, _prevProof);
+      TxLib.Tx memory previousTx = TxLib.parseTx(previousTxData);      
+      uint32 prevHeight = bridge.periods[_prevProof[0]].height;
+
+      require(blockHeight < prevHeight);
+      tokens[transferTx.outs[0].color].addr.transferFrom(address(this), msg.sender, challengeStake);
+
+    }
+    limboExit.isValid = false;
+    limboExit.output[OutputNo].exitable = false;
+    limboExit.challenge = LimboChallenge({
+      owner: msg.sender,
+      resolved: false,
+      proof: _spendProof
+    });
+  }
+
+  function resolveInputSpendChallenge(
+    bytes32 exitId, bytes memory inTxData, uint8 InputNo,
+    bytes32[] memory _txProof, bytes32[] memory _youngestInputProof
+  ) public {
+    LimboExit memory limboExit = limboExits[exitId];
+    LimboChallenge memory challenge = limboExit.input[InputNo].challenge;
+    bytes32[] memory _prevProof = challenge.proof;
+    TxLib.Tx memory transferTx = TxLib.parseTx(inTxData);
+    bytes32 inTxHash = keccak256(inTxData);
+
+    bytes32 previousHash;
+    bytes memory previousTxData;
+    uint64 previousPos;
+
+    (previousPos, previousHash, previousTxData) = TxLib.validateProof(32 * (_youngestInputProof.length + 2) + 64, _prevProof);
+    TxLib.Tx memory previousTx = TxLib.parseTx(previousTxData);      
+    uint32 prevHeight = bridge.periods[_prevProof[0]].height;
+
+    uint32 blockHeight = bridge.periods[_txProof[0]].height;
+    require(blockHeight < prevHeight);
+
+    limboExit.isValid = true;
+    limboExit.input[InputNo].exitable = true;
+    limboExit.input[InputNo].challenge.resolved = true;
+    //[TODO] pay resolver exitStake
+  }
+
+  function finalizeTopLimboExit(uint16 _color) public {
+    bytes32 utxoId;
+    uint256 exitableAt;
+    (utxoId, exitableAt) = getNextExit(_color);
+
+    require(exitableAt <= block.timestamp, "The top exit can not be exited yet");
+    require(tokens[_color].currentSize > 0, "The exit queue for color is empty");
+
+    LimboExit memory currentExit = limboExits[utxoId];
+    uint256 amount;
+    if (currentExit.isValid == true){
+      // all piggybacked outputs are paid out
+      for(uint8 i=0; i < currentExit.output.length; i++){
+        LimboOut memory out = currentExit.output[i];
+        if (out.exitable){
+          amount = piggybackStake + out.amount;
```

Comment:
`piggybackStake` is in wei and `out.amount` is in tokens. Separate transfers needed here